File: https_server.py
#!/usr/bin/env python3
import http.server
import socketserver
import ssl
import os
import subprocess
import sys

# Server configuration
PORT = 8443
DIRECTORY = "."
CERT = 'cert.pem'
KEY = 'key.pem'

def generate_certificates():
    """Generate SSL certificates if they don't exist"""
    if not (os.path.exists(CERT) and os.path.exists(KEY)):
        print('❗ Nema sertifikata! Pokreni sledeću komandu u terminalu:')
        print('openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes -subj "/C=RS/ST=Serbia/L=Belgrade/O=EngleskiApp/CN=localhost"')
        exit(1)

def create_simple_certificates():
    """Create simple self-signed certificates using Python"""
    from cryptography import x509
    from cryptography.x509.oid import NameOID
    from cryptography.hazmat.primitives import hashes, serialization
    from cryptography.hazmat.primitives.asymmetric import rsa
    from datetime import datetime, timedelta
    
    # Generate private key
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
    )
    
    # Create certificate
    subject = issuer = x509.Name([
        x509.NameAttribute(NameOID.COUNTRY_NAME, "RS"),
        x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, "Serbia"),
        x509.NameAttribute(NameOID.LOCALITY_NAME, "Belgrade"),
        x509.NameAttribute(NameOID.ORGANIZATION_NAME, "EngleskiApp"),
        x509.NameAttribute(NameOID.COMMON_NAME, "localhost"),
    ])
    
    cert = x509.CertificateBuilder().subject_name(
        subject
    ).issuer_name(
        issuer
    ).public_key(
        private_key.public_key()
    ).serial_number(
        x509.random_serial_number()
    ).not_valid_before(
        datetime.utcnow()
    ).not_valid_after(
        datetime.utcnow() + timedelta(days=365)
    ).add_extension(
        x509.SubjectAlternativeName([
            x509.DNSName("localhost"),
            x509.IPAddress("127.0.0.1"),
        ]),
        critical=False,
    ).sign(private_key, hashes.SHA256())
    
    # Write certificates to files
    with open(CERT, "wb") as f:
        f.write(cert.public_bytes(serialization.Encoding.PEM))
    
    with open(KEY, "wb") as f:
        f.write(private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption()
        ))
    
    print("✅ SSL sertifikati kreirani pomoću Python-a")

class MyHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        # Add CORS headers for better compatibility
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(200)
        self.end_headers()

if __name__ == "__main__":
    # Change to the directory containing the files
    os.chdir(DIRECTORY)
    
    # Generate certificates if needed
    generate_certificates()
    
    # Create HTTP server
    httpd = socketserver.TCPServer(('0.0.0.0', PORT), MyHTTPRequestHandler)
    
    # Wrap with SSL
    httpd.socket = ssl.wrap_socket(httpd.socket, keyfile=KEY, certfile=CERT, server_side=True)
    
    print(f"🚀 HTTPS Server pokrenut na https://0.0.0.0:{PORT}")
    print(f"📱 Lokalna mreža: https://192.168.31.214:{PORT}")
    print(f"💻 Lokalno: https://localhost:{PORT}")
    print("⚠️  Browser će upozoriti o self-signed sertifikatu - kliknite 'Advanced' -> 'Proceed'")
    print("🛑 Za zaustavljanje pritisnite Ctrl+C")
    
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n🛑 Server zaustavljen")
        httpd.shutdown() 

File: https_server.js
const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const PORT = 8443;
const HTTP_PORT = 8000;

// Function to create self-signed certificate
function createCertificate() {
    if (!fs.existsSync('cert.pem') || !fs.existsSync('key.pem')) {
        console.log('🔐 Kreiranje SSL sertifikata...');
        try {
            // Create certificate using openssl
            const cmd = `openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes -subj "/C=RS/ST=Serbia/L=Belgrade/O=EngleskiApp/CN=localhost"`;
            execSync(cmd, { stdio: 'pipe' });
            console.log('✅ SSL sertifikati uspešno kreirani');
            return true;
        } catch (error) {
            console.log('❌ OpenSSL nije dostupan. Pokretanje HTTP servera...');
            return false;
        }
    }
    return true;
}

// Function to get MIME type
function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const mimeTypes = {
        '.html': 'text/html',
        '.js': 'application/javascript',
        '.css': 'text/css',
        '.json': 'application/json',
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.gif': 'image/gif',
        '.svg': 'image/svg+xml',
        '.ico': 'image/x-icon'
    };
    return mimeTypes[ext] || 'application/octet-stream';
}

// Function to serve files
function serveFile(req, res) {
    let filePath = '.' + req.url;
    
    // Default to index.html
    if (filePath === './') {
        filePath = './index.html';
    }
    
    // Remove query parameters
    filePath = filePath.split('?')[0];
    
    fs.readFile(filePath, (err, data) => {
        if (err) {
            res.writeHead(404, { 'Content-Type': 'text/html' });
            res.end('<h1>404 - File Not Found</h1>');
            return;
        }
        
        const mimeType = getMimeType(filePath);
        res.writeHead(200, { 
            'Content-Type': mimeType,
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type'
        });
        res.end(data);
    });
}

// Create server
if (createCertificate()) {
    // HTTPS Server
    const options = {
        key: fs.readFileSync('key.pem'),
        cert: fs.readFileSync('cert.pem')
    };
    
    const server = https.createServer(options, (req, res) => {
        serveFile(req, res);
    });
    
    server.listen(PORT, '0.0.0.0', () => {
        console.log(`🚀 HTTPS Server pokrenut na https://0.0.0.0:${PORT}`);
        console.log(`📱 Lokalna mreža: https://192.168.31.214:${PORT}`);
        console.log(`💻 Lokalno: https://localhost:${PORT}`);
        console.log('⚠️  Browser će upozoriti o self-signed sertifikatu');
        console.log('   Kliknite "Advanced" -> "Proceed to localhost (unsafe)"');
        console.log('🛑 Za zaustavljanje pritisnite Ctrl+C');
    });
} else {
    // HTTP Server (fallback)
    const server = http.createServer((req, res) => {
        serveFile(req, res);
    });
    
    server.listen(HTTP_PORT, '0.0.0.0', () => {
        console.log(`🚀 HTTP Server pokrenut na http://0.0.0.0:${HTTP_PORT}`);
        console.log(`📱 Lokalna mreža: http://192.168.31.214:${HTTP_PORT}`);
        console.log(`💻 Lokalno: http://localhost:${HTTP_PORT}`);
        console.log('🛑 Za zaustavljanje pritisnite Ctrl+C');
    });
} 